upb/util/def_to_proto.c

```diff
@@ -0,0 +1,507 @@
+    msgs[i] = msgdef_toproto(ctx, upb_filedef_toplvlmsg(f, i));
+  }
+
+  n = upb_filedef_toplvlenumcount(f);
+  google_protobuf_EnumDescriptorProto **enums =
+      google_protobuf_FileDescriptorProto_resize_enum_type(proto, n,
+                                                           ctx->arena);
+  for (int i = 0; i < n; i++) {
+    enums[i] = enumdef_toproto(ctx, upb_filedef_toplvlenum(f, i));
+  }
+
+  n = upb_filedef_servicecount(f);
+  google_protobuf_ServiceDescriptorProto **services =
+      google_protobuf_FileDescriptorProto_resize_service(proto, n, ctx->arena);
+  for (int i = 0; i < n; i++) {
+    services[i] = servicedef_toproto(ctx, upb_filedef_service(f, i));
+  }
+
+  n = upb_filedef_toplvlextcount(f);
+  google_protobuf_FieldDescriptorProto **exts =
+      google_protobuf_FileDescriptorProto_resize_extension(proto, n,
+                                                           ctx->arena);
+  for (int i = 0; i < n; i++) {
+    exts[i] = fielddef_toproto(ctx, upb_filedef_toplvlext(f, i));
+  }
+
+  if (upb_filedef_hasoptions(f)) {
+    SET_OPTIONS(proto, FileDescriptorProto, FileOptions,
+                upb_filedef_options(f));
+  }
+
+  return proto;
+}
+
+google_protobuf_DescriptorProto *upb_MessageDef_ToProto(const upb_msgdef *m,
+                                                        upb_arena *a) {
+  toproto_ctx ctx = {a};
+  return UPB_SETJMP(ctx.err) ? NULL : msgdef_toproto(&ctx, m);
+}
+
+google_protobuf_EnumDescriptorProto *upb_EnumDef_ToProto(const upb_enumdef *e,
+                                                         upb_arena *a) {
+  toproto_ctx ctx = {a};
+  return UPB_SETJMP(ctx.err) ? NULL : enumdef_toproto(&ctx, e);
+}
+
+google_protobuf_EnumValueDescriptorProto *upb_EnumValueDef_ToProto(
+    const upb_enumvaldef *e, upb_arena *a) {
+  toproto_ctx ctx = {a};
+  return UPB_SETJMP(ctx.err) ? NULL : enumvaldef_toproto(&ctx, e);
+}
+
+google_protobuf_FieldDescriptorProto *upb_FieldDef_ToProto(
+    const upb_fielddef *f, upb_arena *a) {
+  toproto_ctx ctx = {a};
+  return UPB_SETJMP(ctx.err) ? NULL : fielddef_toproto(&ctx, f);
+}
+
+google_protobuf_OneofDescriptorProto *upb_OneofDef_ToProto(
+    const upb_oneofdef *o, upb_arena *a) {
+  toproto_ctx ctx = {a};
+  return UPB_SETJMP(ctx.err) ? NULL : oneofdef_toproto(&ctx, o);
+}
+
+google_protobuf_FileDescriptorProto *upb_FileDef_ToProto(const upb_filedef *f,
+                                                         upb_arena *a) {
+  toproto_ctx ctx = {a};
+  return UPB_SETJMP(ctx.err) ? NULL : filedef_toproto(&ctx, f);
+}
+
+google_protobuf_MethodDescriptorProto *upb_MethodDef_ToProto(
+    const upb_methoddef *m, upb_arena *a) {
+  toproto_ctx ctx = {a};
+  return UPB_SETJMP(ctx.err) ? NULL : methoddef_toproto(&ctx, m);
+}
+
+google_protobuf_ServiceDescriptorProto *upb_ServiceDef_ToProto(
+    const upb_servicedef *s, upb_arena *a) {
+  toproto_ctx ctx = {a};
+  return UPB_SETJMP(ctx.err) ? NULL : servicedef_toproto(&ctx, s);
```

Comment:
macros in ternaries wig me out a bit....

Reply:
Done.